pkg/router/template/util/rewritetarget/rewritetarget.go

```diff
@@ -0,0 +1,156 @@
+func removeSingleBackslashes(val string) string {
+	return oneOrMoreBackslashes.ReplaceAllStringFunc(val, func(match string) string {
+		if len(match) == 1 {
+			return oneOrMoreBackslashes.ReplaceAllString(match, "")
+		}
+		return match
+	})
+}
+
+// convertDoubleBackslashes converts double backslashes \\ to single
+// backslashes \.
+func convertDoubleBackslashes(val string) string {
+	return strings.ReplaceAll(val, `\\`, `\`)
+}
+
+// processHashCreator creates a function to process '#' runes in a
+// string. It returns a processFunc which will set
+// encounteredCommentMarker to true and stop processing if it
+// encounters a '#' that is not escaped.
+func processHashCreator(encounteredCommentMarker *bool) processFunc {
+	return func(char rune, escaped bool) runeResult {
+		if char == '#' && !escaped {
+			*encounteredCommentMarker = true
+			return runeResult{stop: true}
+		}
+		return runeResult{value: string(char), escaped: char == '\\' && !escaped}
+	}
+}
+
+// processDoubleQuotes is a processFunc that processes double quote
+// characters. It skips the double quote if it's not escaped.
+var processDoubleQuotes = func(char rune, escaped bool) runeResult {
+	if char == '"' && !escaped {
+		return runeResult{skip: true}
+	}
+	return runeResult{value: string(char), escaped: char == '\\' && !escaped}
+}
+
+// processSingleQuotes is a processFunc that processes single quote
+// characters. It skips the single quote if it's not escaped. If it's
+// escaped, it adds an escaped single quote for later conversion.
+var processSingleQuotes = func(char rune, escaped bool) runeResult {
+	if char == '\'' && !escaped {
+		return runeResult{skip: true}
+	} else if char == '\'' && escaped {
+		// TRICKY: Because our later operation removes single \,
+		// we should double escape \\ it, so it will be converted
+		// to \ later.
+		return runeResult{value: `'\\''`}
+	}
+	return runeResult{value: string(char), escaped: char == '\\' && !escaped}
+}
+
+// SanitizeInput processes the `haproxy.router.openshift.io/rewrite-target`
+// annotation value for API compatibility while properly handling values
+// with spaces, backslashes, and other special characters. Because the
+// annotation value was initially introduced without being enclosed in
+// single quotes, certain characters were interpreted rather than being
+// treated literally. We later added the single quotes wrapping to resolve
+// OCPBUGS-22739. However, we must still maintain API compatibility after
+// this change: the annotation values MUST be interpreted to the same values
+// after updating to enclose the value in single quotes.
+func SanitizeInput(val string) string {
+	var encounteredCommentMarker bool
+
+	val = processRunes(val, processHashCreator(&encounteredCommentMarker))
+	val = processRunes(val, processDoubleQuotes)
+	val = processRunes(val, processSingleQuotes)
+	val = replacePercentSigns(val)
+	val = removeSingleBackslashes(val)
+	val = convertDoubleBackslashes(val)
```

Comment:
Given that there's differing function signatures you would have to do something like:

```go
func SanitizeInput(val string) string {
	type processStep func(string) string

	var encounteredCommentMarker bool

	fns := []processStep{
		func(val string) string { return processRunes(val, processHashCreator(&encounteredCommentMarker)) },
		func(val string) string { return processRunes(val, processDoubleQuotes) },
		func(val string) string { return processRunes(val, processSingleQuotes) },
		replacePercentSigns,
		removeSingleBackslashes,
		convertDoubleBackslashes,
	}

	for _, fn := range fns {
		val = fn(val)
	}

	if !encounteredCommentMarker {
		// The literal `\1` is appended to annotations without
		// comments to comply with HAProxy rewrite rule
		// expectations, which necessitate a capture group
		// reference (i.e., `\1`) for dynamic substitutions.
		val += `\1`
	}

	return val
}
```

Is this actually any better than just calling known functions names one by one?

Reply:
> Is this actually any better than just calling known functions names one by one?

I find it slightly more structured. But I agree that this may be subjective. So, I'm fine to drop it.

Reply:
Sounds good, leaving as is.